File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    //function for finding the prefix product 
    void preProd(vector<int>&ans ,vector<int>arr ){
        int n = arr.size(); 
        int prod =1 ;
        for (int i=0 ;i<n ;i++ ){
            ans[i] = prod ; 
            prod *= arr[i];
        }
        return ;

    }

     void suffProd(vector<int>&ans ,vector<int>arr ){
        int n = arr.size(); 
        int prod =1 ;
        for (int i=n-1 ;i>=0 ;i-- ){
            ans[i] = prod ; 
            prod *= arr[i];
        }
        return ;

    }
    vector<int> productExceptSelf(vector<int>& nums) {
        // concept : prefix product and suffix product ; 
            int n = nums.size(); 
        vector<int> ans(n,0);
        vector<int>prefix(n,0); 
        preProd(prefix,nums);
        vector<int>suffix(n,0);
        suffProd(suffix,nums);

        for (int i=0 ;i<n;i++ ){
            ans[i]  = (prefix[i]* suffix[i]);
        }
        return ans ; 
    

    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    // prefix Product and suffix Product concept will be used 
    let n = nums.length;
    let pref=1 , suff =1 ;
    let prefixProd = new  Array(n);; 
    let suffixProd = Array(n);
    let ans=  new Array(n);


    prefixProd[0] =1 ; 
    suffixProd[n-1] = 1;
    
    for(let i= 1 ;i<n ;i++ ){
        pref *= nums[i-1];
        prefixProd[i] = pref;
    }
    for (let i=n-2 ;i>=0 ;i-- ){
        suff *=nums[i+1];
        suffixProd[i] = suff ;
    }

    // actual calculation 
    for (let i=0 ;i<n ;i++ ){
        nums[i] = (prefixProd[i] * suffixProd[i]);
    }
    return nums;

};